feature(ivy): record absolute position of template expressions

ayazhafiz opened the pull request:
Currently, template expressions and statements have their location
recorded relative to the HTML element they are in, with no handle to
absolute location in a source file except for a line/column location.
However, the line/column location is also not entirely accurate, as it
points an entire semantic expression, and not necessarily the start of
an expression recorded by the expression parser.

To support record of the source code expressions originate from, add a
new `sourceSpan` field to `ASTWithSource` that records the absolute byte
offset of an expression within a source code.

Implement part 2 of [refactoring template parsing for
stability](https://hackmd.io/@X3ECPVy-RCuVfba-pnvIpw/BkDUxaW84/%2FMA1oxh6jRXqSmZBcLfYdyw?type=book).

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?

Currently, template expressions and statements have their location
recorded relative to the HTML element they are in, with no handle to
absolute location in a source file except for a line/column location.
However, the line/column location is also not entirely accurate, as it
points an entire semantic expression, and not necessarily the start of
an expression recorded by the expression parser.

## What is the new behavior?

To support record of the source code expressions originate from, add a
new `sourceSpan` field to `ASTWithSource` that records the absolute byte
offset of an expression within a source code.

## Does this PR introduce a breaking change?

- [x] Yes
- [ ] No

## Other information

See the [design doc for refactoring template parsing for stability](https://hackmd.io/@X3ECPVy-RCuVfba-pnvIpw/BkDUxaW84/%2FMA1oxh6jRXqSmZBcLfYdyw?type=book).

ayazhafiz commented:
Depends on #31347. Will rebase once merged.

JoostK commented on packages/compiler/src/template_parser/binding_parser.ts:
```diff
@@ -49,14 +49,16 @@ export class BindingParser {
-      ParsedProperty[]|null {
+  createBoundHostProperties(
+      dirMeta: CompileDirectiveSummary, sourceSpan: ParseSourceSpan,
+      absoluteOffset: number = 0): ParsedProperty[]|null {
```
What is the reasoning behind making `absoluteOffset` optional here? I'm concerned it will be too easy not to pass any value here, leading to inaccurate offset data. WDYT?

ayazhafiz replied:
I wanted to change as little existing code as possible. I think a more reasonable default is `sourceSpan.start.offset`, since that's what users of this method would probably pass anyway. Updated to use that as a default.

JoostK replied:
Works for me, thanks for elaborating!

JoostK commented on packages/compiler/src/template_parser/binding_parser.ts:
```diff
@@ -69,9 +71,9 @@ export class BindingParser {
-      sourceSpan: ParseSourceSpan): BoundElementProperty[]|null {
-    const boundProps = this.createBoundHostProperties(dirMeta, sourceSpan);
+      dirMeta: CompileDirectiveSummary, elementSelector: string, sourceSpan: ParseSourceSpan,
+      absoluteOffset: number = 0): BoundElementProperty[]|null {
```
Same here with the optional param.

ayazhafiz replied:
See conversation above.

JoostK commented on packages/compiler/test/template_parser/template_parser_spec.ts:
```diff
@@ -1082,6 +1128,35 @@ Binding to attribute 'onEvent' is disallowed for security reasons ("<my-componen
 
       });
 
+      describe('offset', () => {
```
I have a feeling that `TemplateParser` is mostly obsolete once Ivy is done, in particular because in Ivy's compiler the binding of directives etc. is done using `TargetBinder` and there's a dedicated R3 AST. Therefore, these new testcases might fit in better in the new infrastructure. You can use https://github.com/angular/angular/blob/673ac2945c253b2b1d5947b7c5570a3171be7d89/packages/compiler/test/render3/view/util.ts#L80-L81 as an alternative to the `parse` function in this suite of tests. I have a similar test for error mapping here: https://github.com/angular/angular/blob/2440513b26546cd6b63f53bd8ae79a48653f28ec/packages/compiler/test/render3/r3_ast_spans_spec.ts

@alxhub WDYT?

alxhub replied:
Agreed.

I'm not against having the tests here, but I do think they would be better served as tests on the Ivy side, since VE won't read these fields at all.

ayazhafiz replied:
Moved tests under render3.

alxhub reviewed (commented):
Also make this a `feat(ivy)` as there's no need to announce this commit to the user, especially since it won't be useful in VE at all.

ayazhafiz committed:
feat(ivy): record absolute position of template expressions

Currently, template expressions and statements have their location
recorded relative to the HTML element they are in, with no handle to
absolute location in a source file except for a line/column location.
However, the line/column location is also not entirely accurate, as it
points an entire semantic expression, and not necessarily the start of
an expression recorded by the expression parser.

To support record of the source code expressions originate from, add a
new `sourceSpan` field to `ASTWithSource` that records the absolute byte
offset of an expression within a source code.

Implement part 2 of [refactoring template parsing for
stability](https://hackmd.io/@X3ECPVy-RCuVfba-pnvIpw/BkDUxaW84/%2FMA1oxh6jRXqSmZBcLfYdyw?type=book).

ayazhafiz committed:
fixup! feat(ivy): record absolute position of template expressions

ayazhafiz commented:
@JoostK, @alxhub PTAL with updated commits.

JoostK commented on packages/compiler/src/template_parser/binding_parser.ts:
```diff
@@ -49,14 +49,24 @@ export class BindingParser {
+  createBoundHostProperties(
+      dirMeta: CompileDirectiveSummary, sourceSpan: ParseSourceSpan,
+      absoluteOffset?: number): ParsedProperty[]|null {
+    if (!absoluteOffset) {
```
```suggestion
    if (absoluteOffset === undefined) {
```

Especially because `absoluteOffset` is of type `number`, so we shouldn't mess with 0.

JoostK commented on packages/compiler/src/template_parser/binding_parser.ts:
```diff
@@ -49,14 +49,24 @@ export class BindingParser {
+      // sourceSpan. Perfom this check until that is fixed.
+      absoluteOffset = sourceSpan ? sourceSpan.start.offset : 0;
+    }
+    // Typechecker can't deduce that `absoluteOffset` is a number in the conditional blocks below,
```
That's weird! Even if the above guard uses `=== undefined`?

ayazhafiz replied:
Yeah, here's a simpler example: [playground](http://www.typescriptlang.org/play/#code/DYUwLgBAhgXAdgVwLYCMQCcA+C4BMQBmAlnCLgNwBQRBEAFFBALwsQ77Gm4CU0zEAJiqUA2gEYANAIkBmALoA6AgHt0AUSgBjABZ04zAHwQA3pQjmIm5XADOkFDAiJUGflCoBfbuSA).

I think this may be related to https://github.com/microsoft/TypeScript/issues/11498

JoostK replied:
Oooh, there's a closure involved. Then it makes sense :-)

JoostK commented on packages/compiler/test/render3/r3_ast_absolute_span_spec.ts:
```diff
@@ -0,0 +1,72 @@
+describe('expression AST absolute source spans', () => {
+  it('should provide absolute offsets with arbitrary whitespace', () => {
+    expect(humanizeExpressionLocation(
+               parse('<div>\n  \n{{foo}}</div>', {preserveWhitespaces: true}).nodes))
```
Will setting `preserveWhitespaces` to `false` cause this test to fail, at this moment? I presume so, and that fixing it is in scope for [part 4 of the plan](https://hackmd.io/@X3ECPVy-RCuVfba-pnvIpw/BkDUxaW84/%2FMA1oxh6jRXqSmZBcLfYdyw?type=book), maybe we should include a test to extend the tests once that is done?

ayazhafiz replied:
What do you mean by a test to extend the tests?

JoostK replied:
Sorry, I meant to type TODO! We should consider adding a TODO to extend the tests by then. Sorry again for the confusion.

ayazhafiz replied:
No worries, fixed.

ayazhafiz committed:
fixup! feat(ivy): record absolute position of template expressions

ayazhafiz committed:
fixup! feat(ivy): record absolute position of template expressions

ayazhafiz committed:
fixup! feat(ivy): record absolute position of template expressions

kyliau reviewed (approved):
+1 for language-service

ayazhafiz committed:
fixup! feat(ivy): record absolute position of template expressions

ayazhafiz committed:
fixup! feat(ivy): record absolute position of template expressions

ayazhafiz commented on packages/compiler/src/expression_parser/parser.ts:
```diff
@@ -35,42 +35,46 @@ export class Parser {
       input: string, location: any,
-      interpolationConfig: InterpolationConfig = DEFAULT_INTERPOLATION_CONFIG): ASTWithSource {
+      interpolationConfig: InterpolationConfig = DEFAULT_INTERPOLATION_CONFIG,
+      absoluteOffset: number = 0): ASTWithSource {
```
The problem is that it's a breaking change if we do that; some code in g3 relies on the parameter ordere here.

ayazhafiz replied:
Fixed by patching `codelyzer` dependency on this.

ayazhafiz commented on packages/compiler/src/expression_parser/ast.ts:
```diff
@@ -203,11 +203,17 @@ export class FunctionCall extends AST {
   }
 }
 
+export class AbsoluteSourceSpan {
```
Done.

ayazhafiz commented on packages/compiler/src/expression_parser/parser.ts:
```diff
@@ -35,42 +35,46 @@ export class Parser {
-    const ast = this._parseBindingAst(input, location, interpolationConfig);
-    return new ASTWithSource(ast, input, location, this.errors);
+      interpolationConfig: InterpolationConfig = DEFAULT_INTERPOLATION_CONFIG,
+      absoluteOffset: number = 0): ASTWithSource {
```
See above.

ayazhafiz commented on packages/compiler/src/expression_parser/parser.ts:
```diff
@@ -228,9 +235,9 @@ export class _ParseAST {
-      public inputLength: number, public parseAction: boolean, private errors: ParserError[],
-      private offset: number) {}
+      public input: string, public location: any, public absoluteOffset: number,
+      public tokens: Token[], public inputLength: number, public parseAction: boolean,
```
It just moved onto the next line due to formatting changes caused by the new `absoluteOffset` paramter.

ayazhafiz commented on packages/compiler/src/template_parser/binding_parser.ts:
```diff
@@ -49,14 +49,24 @@ export class BindingParser {
-      ParsedProperty[]|null {
+  createBoundHostProperties(
+      dirMeta: CompileDirectiveSummary, sourceSpan: ParseSourceSpan,
+      absoluteOffset?: number): ParsedProperty[]|null {
```
To minimize changes throughout the entire compiler, I want to rely on a default value of the starting `sourceSpan` byte offset (per the discussion in the lines below).

ayazhafiz replied:
Fixed by setting default value to sourceSpan start offset.

ayazhafiz committed:
fixup! feat(ivy): record absolute position of template expressions

ayazhafiz commented on packages/compiler/src/template_parser/binding_parser.ts:
```diff
@@ -49,14 +49,24 @@ export class BindingParser {
+    if (absoluteOffset === undefined) {
+      // TODO(ayazhafiz): Apparently some consumers of this method are not passing a valid a
+      // sourceSpan. Perfom this check until that is fixed.
+      absoluteOffset = sourceSpan ? sourceSpan.start.offset : 0;
```
I'm going to try and see if there's an easy fix for whatever is passing an invalid sourceSpan so we can get rid of this entirely.

ayazhafiz replied:
Okay, fixed by passing incorrect by valid source span to a caller that was passing a null source span.

ayazhafiz committed:
fixup! feat(ivy): record absolute position of template expressions

ayazhafiz committed:
fixup! feat(ivy): record absolute position of template expressions

ayazhafiz committed:
fixup! feat(ivy): record absolute position of template expressions

ayazhafiz committed:
fixup! feat(ivy): record absolute position of template expressions

ayazhafiz committed:
fixup! feat(ivy): record absolute position of template expressions

alxhub commented on packages/compiler/src/template_parser/binding_parser.ts:
```diff
@@ -49,14 +49,16 @@ export class BindingParser {
-      ParsedProperty[]|null {
+  createBoundHostProperties(
+      dirMeta: CompileDirectiveSummary, sourceSpan: ParseSourceSpan,
+      absoluteOffset: number = sourceSpan.start.offset): ParsedProperty[]|null {
```
Why is this a sensible default value? Under what circumstances will this be correct/not correct?

ayazhafiz replied:
Because in this case, `sourceSpan` is a span of byte offsets in the source file, which is what we want from `absoluteOffset` (`sourceSpan` is later used to compute the line/column offset for error locations).

This is incorrect in places where users of the function pass arbitrary or incorrect values of `sourceSpan`, which unfortunately there is a fair amount of.

ayazhafiz commented on packages/compiler-cli/src/ngtsc/annotations/src/directive.ts:
```diff
@@ -6,7 +6,8 @@
 
-import {ConstantPool, Expression, ParseError, ParsedHostBindings, R3DirectiveMetadata, R3QueryMetadata, Statement, WrappedNodeExpr, compileDirectiveFromMetadata, makeBindingParser, parseHostBindings, verifyHostBindings} from '@angular/compiler';
+import {ConstantPool, Expression, ParseError, ParseSourceSpan, ParsedHostBindings, R3DirectiveMetadata, R3QueryMetadata, Statement, WrappedNodeExpr, compileDirectiveFromMetadata, makeBindingParser, parseHostBindings, verifyHostBindings} from '@angular/compiler';
+import {ParseLocation, ParseSourceFile} from '@angular/compiler/src/compiler';
```
Fixed.

ayazhafiz committed:
fixup! feat(ivy): record absolute position of template expressions

ayazhafiz committed:
fixup! feat(ivy): record absolute position of template expressions

ayazhafiz commented:
@[caretaker]: This PR requires http://cl/258819283 to be patched.

alxhub commented on packages/compiler-cli/src/ngtsc/annotations/src/base_def.ts:
```diff
@@ -91,10 +91,12 @@ export class BaseDefDecoratorHandler implements
 
   analyze(node: ClassDeclaration, metadata: R3BaseRefDecoratorDetection):
       AnalysisOutput<R3BaseRefMetaData> {
+    const EMPTY_PARSE_LOCATION = new ParseLocation(new ParseSourceFile('', ''), 0, 0, 0);
```
Can this be an outer constant so we're not creating one for every single component? Maybe export a constant in a standard location next to where `ParseLocation`/`ParseSourceFile`/`ParseSpan` are defined?

ayazhafiz replied:
Yes!

ayazhafiz committed:
fixup! feat(ivy): record absolute position of template expressions

ayazhafiz commented:
Caretaker: This PR requires http://cl/258819283 to be patched.
